nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -611,21 +614,148 @@ def permuted_ols(
+        intercept_test,
+        n_samples=tested_vars.shape[0],
+    )
+
+    # OLS regression on original data
+    # original regression (= regression on residuals + adjust t-score)
+    # compute t score map of each tested var for original data
+    # scores_original_data is in samples-by-regressors shape
+    confounding_vars = _orthonormalize_confounding_vars(confounding_vars)
+    target_vars = _prepare_target_vars(target_vars, confounding_vars)
+    tested_vars = _prepare_tested_vars(tested_vars, confounding_vars)
+    scores_original_data = _t_score_with_covars_and_normalized_design(
+        tested_vars,
+        target_vars.T,
+        confounding_vars,
+    )
+
+    # Define connectivity for TFCE and/or cluster measures
+    tfce_original_data = _get_tfce_original_data(
+        tfce, masker, scores_original_data, two_sided_test
+    )
+
+    # 0 or negative number of permutations => original data scores only
+    if n_perm <= 0:
+        if output_type == "legacy":
+            return np.asarray([]), scores_original_data.T, np.asarray([])
+
+        out = {"t": scores_original_data.T}
+
+        if tfce:
+            out["tfce"] = tfce_original_data.T
+
+        return out
+
+    # Permutations
+    # parallel computing units perform a reduced number of permutations each
+    return _run_permutations(
+        tested_vars,
+        target_vars,
+        confounding_vars,
+        n_perm,
+        random_state,
+        two_sided_test,
+        n_jobs,
+        verbose,
+        masker,
+        tfce,
+        threshold,
+        output_type,
+        n_descriptors,
+        intercept_test,
+        scores_original_data,
+        tfce_original_data,
+    )
+
+
+def _check_args_permuted_ols(target_vars, n_jobs, tfce, threshold, masker):
+    """Validate input arguments for permuted_ols.
+
+    Parameters
+    ----------
+    target_vars : array-like, shape=(n_samples, n_descriptors)
+        fMRI data to analyze according to the explanatory and confounding
+        variates.
+
+    n_jobs : :obj:`int`
+        Number of parallel workers.
+        If -1 is provided, all CPUs are used.
+        A negative number indicates that all the CPUs except (abs(n_jobs) - 1)
+
+    tfce : :obj:`bool`
+        Whether to calculate :term:`TFCE` as part of the permutation procedure
+        or not.
+        The TFCE calculation is implemented as described in
+        :footcite:t:`Smith2009a`.
+
+    threshold : None or :obj:`float`
+        Cluster-forming threshold in p-scale.
+        This is only used for cluster-level inference.
+        If None, cluster-level inference will not be performed.
+
+    masker : None or :class:`~nilearn.maskers.NiftiMasker` or \
+            :class:`~nilearn.maskers.MultiNiftiMasker`
+        A mask to be used on the data.
+        This is required for cluster-level inference, so it must be provided
+        if ``threshold`` is not None.
+
+    """
+    if target_vars.ndim != 2:
+        raise ValueError(
```

Comment:
If the coverage values are not handled in this PR, they should be handled later.

Reply:
Refactoring without enough coverage is to software development is what free soloing is to climbing.
Will improve coverage.
Ideally and where applicable I will create the test on main and then make sure that they still pass after refactoring to make sure nothing broke.

Reply:
added a test